### PR TITLE
fix: `PrintDialog` not shown

### LINKED ISSUE #2814: Invoke PrintDialog.ShowDialog() does not show Print dialog successfully
* .NET Core Version: .NET Core SDK 5.0.100-preview.1.20106.1 + the VSIX Inserted into Dev16 16.5.0 [29803.141.master]

* Have you experienced this same bug with .NET Framework?: No

**Problem description:**
 Invoke PrintDialog.ShowDialog() does not show Print dialog successfully.

**More Info:**
1. This is a regression issue, it's not reproducing when targeting .NET Core 3.1.
2. This issue also cannot reproduce with Framework project.

**Actual behavior:**
No expected Print dialog shown.
.NET Core 5.0 screenshot:
PrintDialogIssue5 0

**Expected behavior:**
The Print dialog should be shown successfully.
.NET Core 3.1 screenshot:
PrintDialogIssue3 1

**Minimal repro:**

1. Create a Winforms Core application.
2. Drag and drop a button to the form.
3. Double click the button1 to generate the button1_click event, then add following code to it:
PrintDialog pdlg = new PrintDialog();
pdlg.ShowDialog();
4. Build and run.
5. Click the button1 to see whether the Print dialog is popping up.
[PrintDialogIssue.zip](https://github.com/dotnet/winforms/files/4169604/PrintDialogIssue.zip)

### PR DESCRIPTION
**Pull request by RussKie**
Resolves #2814

## Proposed changes

The bug was caused by the conversion of a struct fields to auto-properties in  #2594. This broke the interop binary layout.

Rework the code to
- correctly define x86 and x64 structs
- make the structs blittable
- add tests verifying the structs layouts

## Customer Impact

- `PrintDialog` is now shown as expected
- 

## Regression? 

- Yes

## Test methodology 

manual

* x86
```powershell
dotnet restore --runtime win-x86
dotnet run --runtime win-x86
```
image

* x64
```powershell
dotnet restore --runtime win-x64
dotnet run --runtime win-x64
```
image

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2851)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs [hunk 1]
```diff
@@ -242,150 +242,28 @@ public class EDITSTREAM64
...
+            public Comdlg32.PD Flags;
+            public ushort nFromPage;
+            public ushort nToPage;
+            public ushort nMinPage;
+            public ushort nMaxPage;
+            public ushort nCopies;
+            public IntPtr hInstance;
+            public IntPtr lCustData;
+            public WndProc lpfnPrintHook;
+            public WndProc lpfnSetupHook;
```
Comment (gpetrou) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: Maybe you could use code from Interop.WNDPROC.cs here?

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs [hunk 2]
```diff
@@ -242,150 +242,28 @@ public class EDITSTREAM64
...
+            public IntPtr hDevMode;
+            public IntPtr hDevNames;
+            public IntPtr hDC;
+            public Comdlg32.PD Flags;
+            public ushort nFromPage;
+            public ushort nToPage;
+            public ushort nMinPage;
+            public ushort nMaxPage;
+            public ushort nCopies;
+            public IntPtr hInstance;
```
Comment (weltkante) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: By no longer having separate structs for 32bit/64bit you are breaking the layout here, 32 bit will add 2 byte padding here where none is expected.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs [hunk 3]
```diff
@@ -242,150 +242,28 @@ public class EDITSTREAM64
...
-            public IntPtr hPrintTemplate { get { return m_hPrintTemplate; } set { m_hPrintTemplate = value; } }
-            public IntPtr hSetupTemplate { get { return m_hSetupTemplate; } set { m_hSetupTemplate = value; } }
-        }
-
-        // Any change in PRINTDLG_64, should also be in PRINTDLG_32 and PRINTDLG
-        // x64 does not require EXPLICIT packing of 1.
-        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Auto)]
-        public class PRINTDLG_64 : PRINTDLG
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        public struct PRINTDLGW
```
Comment (hughbe) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: can we extract these classes to `Interop.Comdlg32`? also the comment above is no longer relevant

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs (hunk 2 above)
Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: Relevant: https://github.com/tpn/winsdk-10/blob/9b69fd26ac0c7d0b83d378dba01080e93349c2ed/Include/10.0.10240.0/um/commdlg.h#L49

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs [hunk 4]
```diff
@@ -242,150 +242,28 @@ public class EDITSTREAM64
...
+            public ushort nToPage;
+            public ushort nMinPage;
+            public ushort nMaxPage;
+            public ushort nCopies;
+            public IntPtr hInstance;
+            public IntPtr lCustData;
+            public WndProc lpfnPrintHook;
+            public WndProc lpfnSetupHook;
+            public string lpPrintTemplateName;
+            public string lpSetupTemplateName;
```
Comment (hughbe) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: nit: if we turn these into `char*` (they are unused), make `WndProc` into `IntPtr` and use `Marshal.GetFunctionPointerForDelegate`, then we can make this blittable

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs (hunk 2 above)
Comment (weltkante) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: Yes, original source also had this in comments/attributes, 32bit uses Pack=1, 64bit uses default packing. Since the headers explicitely switch packing depending on bitness you probably can't avoid having two layouts, at least I can't think of any single definition compatible with both.

Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: Could have #if WIN32 or something?

Comment (weltkante) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: `#if` is a compile-time directive, I don't think you want to compile different C# assemblies for 32/64 bit do you? Usually you only have one anycpu assembly for .NET

**Comment by RussKie** [comment]
Thank you for the reviews. This was a quick EOD POC to see it fixed. More cleanups are in order, as you rightly pointed.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: Thank you for pointing this one out. Very insightful.
I have built an x86 variant, and sure enough, it didn't work.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs (hunk 3 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: done

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs (hunk 4 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: done

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: Ended up changing it an `IntPtr` as per @hughbe's recommendation to make it blittable (which not only made it slightly more optimal, but also enabled tests)

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs [hunk 5]
```diff
@@ -353,26 +319,41 @@ internal static NativeMethods.PRINTDLGEX CreatePRINTDLGEX()
...
+            data.nFromPage = 1;
+            data.nToPage = 1;
+            data.nMinPage = 0;
+            data.nMaxPage = 9999;
             data.Flags = GetFlags();
-            data.nCopies = (short)PrinterSettings.Copies;
+            data.nCopies = (ushort)PrinterSettings.Copies;
             data.hwndOwner = hwndOwner;
-            data.lpfnPrintHook = hookProcPtr;
+            data.lpfnPrintHook = Marshal.GetFunctionPointerForDelegate(new User32.WNDPROCINT(HookProc));
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs]: Is there a more optimal way of doing this?

File: src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/MainForm.Designer.cs [hunk 6]
```diff
@@ -217,6 +218,16 @@ private void InitializeComponent()
             this.ShowExceptionDialogButton.UseVisualStyleBackColor = true;
             this.ShowExceptionDialogButton.Click += new System.EventHandler(this.ShowExceptionDialogButton_Click);
             // 
+            // ShowPrintDialogButton
+            // 
+            this.ShowPrintDialogButton.Location = new System.Drawing.Point(3, 177);
+            this.ShowPrintDialogButton.Name = "ShowPrintDialogButton";
+            this.ShowPrintDialogButton.Size = new System.Drawing.Size(259, 23);
+            this.ShowPrintDialogButton.TabIndex = 6;
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/MainForm.Designer.cs]: I still need to fix the taborder on the form. Or make this the last element... TBD

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/NativeMethods.cs]: I asked internally and there is no desire to build leaf-node stacks per arch, as it is very involved. The binaries get cross-gened for each setup bundle.
This means we must have two different versions of the struct (due to different packing requirements), and jump through hoops to populate the payload and invoke the underlying Win32 API.

File: Directory.Build.targets [hunk 7]
```diff
@@ -17,7 +17,9 @@
   </PropertyGroup>
 
   <PropertyGroup Condition="'$(IsTestProject)' == 'true'">
-    <TestArchitectures>x64;x86</TestArchitectures>
+    <Platforms Condition="'$(Platform)' == 'x86'">AnyCPU;x86</Platforms>
+    <TestArchitectures Condition="'$(Platform)' == 'x86'">x86</TestArchitectures>
+    <TestArchitectures Condition="'$(Platform)' != 'x86'">x64;x86</TestArchitectures>
```
Comment (RussKie) [inline Directory.Build.targets]: This enables us running x86 tests locally from cli:
```
.\build.cmd -platform x86 -test
```


Unfortunately I couldn't get them run from VS, as it would not invoke the 32-bit xUnit runner.

File: src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace System.Windows.Forms.Primitives.Tests.Interop.Comdlg32
+{
+    public class PRINTDLGWTests : IClassFixture<ThreadExceptionFixture>
+    {
+        public static bool Is32bit => IntPtr.Size == 4;
+        public static bool Is64bit => IntPtr.Size == 8;
+
+        [ConditionalFact(nameof(Is32bit))]
+
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs]: extra white line

File: src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        public unsafe void PRINTDLGW_32_Size()
+        {
+            Assert.Equal(66, Marshal.SizeOf<PRINTDLGW_32>());
+        }
+
+        [ConditionalFact(nameof(Is32bit))]
+        public unsafe void PRINTDLGW_32_ensure_layout()
+        {
+            PRINTDLGW_32 sut = new PRINTDLGW_32();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs]: What does `sut` stand for?

File: src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            Assert.Equal(42, (byte*)&sut._lpfnPrintHook - addr);        // 4, LPPRINTHOOKPROC
+            Assert.Equal(46, (byte*)&sut._lpfnSetupHook - addr);        // 4, LPSETUPHOOKPROC
+            Assert.Equal(50, (byte*)&sut._lpPrintTemplateName - addr);  // 4, LPCWSTR
+            Assert.Equal(54, (byte*)&sut._lpSetupTemplateName - addr);  // 4, LPCWSTR
+            Assert.Equal(58, (byte*)&sut._hPrintTemplate - addr);       // 4, HGLOBAL
+            Assert.Equal(62, (byte*)&sut._hSetupTemplate - addr);       // 4, HGLOBAL
+        }
+
+        [ConditionalFact(nameof(Is64bit))]
+
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs]: extra line

File: Directory.Build.targets (hunk 7 above)
Comment (weltkante) [reply Directory.Build.targets]: Changing the bitness from the VS "Test" menu doesn't work?

File: src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs (hunk 9 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs]: In general this stands for "system under test", in this case - "struct under test".
I couldn't come up with a better name. Open for ideas.

File: Directory.Build.targets (hunk 7 above)
Comment (RussKie) [reply Directory.Build.targets]: No, it didn't to my surprise.

* Built x86, directed tests to be in x86, the runner is x64
image
* Built AnyCPU, directed tests to be in x86, the runner is x64
image

Comment (weltkante) [reply Directory.Build.targets]: Seems this is a general bug for .NET Core projects, I can reproduce it in a new test project immediately and when you change e.g. to net472 as TargetFramework it properly selects the right bitness. Can't tell if its caused by VS or xUnit, so not sure where this should be raised for further investigation, but its definitely a general problem and not caused by the WinForms project setup.

Comment (RussKie) [reply Directory.Build.targets]: Thank you for checking 👍 
Do you still the repro project? It'll save me time for the escalation

Comment (weltkante) [reply Directory.Build.targets]: Yes I still have it, though I initially ran my tests in VS 2017 - didn't bother with VS 2019 because behavior matched exactly with what you described. Now that you asked I actually ran it under VS 2019 and behavior is slightly different and no longer matches what you see in WinForms. I'm gonna do some more testing to see why it diverges from you observations and create a github repo if there is something to be reported.

Comment (RussKie) [reply Directory.Build.targets]: The positive side from this - I learned how to run tests from cli :)

File: src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            Assert.Equal(50, (byte*)&sut._lpPrintTemplateName - addr);  // 4, LPCWSTR
+            Assert.Equal(54, (byte*)&sut._lpSetupTemplateName - addr);  // 4, LPCWSTR
+            Assert.Equal(58, (byte*)&sut._hPrintTemplate - addr);       // 4, HGLOBAL
+            Assert.Equal(62, (byte*)&sut._hSetupTemplate - addr);       // 4, HGLOBAL
+        }
+
+        [ConditionalFact(nameof(Is64bit))]
+        public unsafe void PRINTDLGW_64_Size()
+        {
+            Assert.Equal(120, Marshal.SizeOf<PRINTDLGW_64>());
```
Comment (weltkante) [inline src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs]: As noted [here](https://github.com/dotnet/winforms/pull/2911#discussion_r384452969) there are differences between blitting structs and classic marshalling. Your test cases are mixing up both styles so you test the size using classic marshalling and offsets using blitting layout. You may want to be consistent in what you test. (Though as far as I'm currently away it can only differ if `bool` is used somewhere.)

* `Marshal.SizeOf` for classic marshalling size
* `Marshal.OffsetOf` for classic marshalling offsets
* `sizeof` keyword for blitting size
* pointer arithmetic for blitting offsets

Example where it makes a difference:
```csharp
public struct Sample
{
    public int a;
    public bool b;
    public short c;
}

var sample = new Sample();
var blittingSize = sizeof(Sample);                              // 8
var blittingOffset = (byte*)&sample.c - (byte*)&sample;         // 6
var marshalSize = Marshal.SizeOf<Sample>();                     // 12
var marshalOffset = Marshal.OffsetOf<Sample>(nameof(Sample.c)); // 8
```

Comment (RussKie) [reply src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs]: Thank you 👍  
It started as a non-blittable struct and I think that's when I started writing tests, and forgot to switch. Fixed now.

File: src/System.Windows.Forms.Primitives/src/Interop/Comdlg32/Interop.PrintDlg.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            if (IntPtr.Size == 4)
+            {
+                if (lppd is PRINTDLGW_32 lppd32)
+                {
+                    BOOL result = PrintDlgW(ref lppd32);
+                    lppd = lppd32;
+                    return result;
+                }
+
+                throw new InvalidOperationException($"Expected {nameof(PRINTDLGW_32)} data struct");
```
Comment (hughbe) [inline src/System.Windows.Forms.Primitives/src/Interop/Comdlg32/Interop.PrintDlg.cs]: do you think this should be a debug assert and a hard cast?

File: src/System.Windows.Forms.Primitives/src/Interop/Comdlg32/Interop.PrintDlgW.cs [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+            public IntPtr _lCustData;
+            public IntPtr _lpfnPrintHook;
+            public IntPtr _lpfnSetupHook;
+            public char* _lpPrintTemplateName;
+            public char* _lpSetupTemplateName;
+            public IntPtr _hPrintTemplate;
+            public IntPtr _hSetupTemplate;
+
+            public uint lStructSize { get => _lStructSize; set => _lStructSize = value; }
+            public IntPtr hwndOwner { get => _hwndOwner; set => _hwndOwner = value; }
```
Comment (hughbe) [inline src/System.Windows.Forms.Primitives/src/Interop/Comdlg32/Interop.PrintDlgW.cs]: can we convert these into simple fields - this is what we do in other structs

Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/Interop/Comdlg32/Interop.PrintDlgW.cs]: Yes, I had several goes at this, trying to do without the interface and inheritance. The code had to duplicate somewhere, and this way seems the lesser evil...

File: src/System.Windows.Forms.Primitives/src/Interop/Comdlg32/Interop.PrintDlg.cs (hunk 12 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/Interop/Comdlg32/Interop.PrintDlg.cs]: The original code had it like this, I just replaced NRE with IOE which is more appropriate.

How do you think `Debug.Assert` will help?
I don't think our code can ever be in a situation to reach this point. We're forced to write the code this way because the input parameter is a ref type (interface) but the actual data is a struct.
To get here an app developer would have do something very dodgy... At that point there won't be any `Debug.Asserts` at the runtime level.

**Comment by RussKie** [comment]
@msftbot merge in 24 hours

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs (hunk 5 above)
Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs]: nothing I'm aware of - you can cache the function pointer in a static variable if the hook is a static method, but then you also need to cache the delegate to prevent it being collected. Overall I don't think its worth it, showing print dialogs isn't exactly a hot loop.

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs]: actually I'm wondering if this needs a GC.KeepAlive on the delegate. There is code running before the dialog is shown so there could be a small risk the GC collects the delegate before entering native code which uses the hook?

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs]: couldn't get it to crash by inserting explicit GC.Collect in the code, so its probably fine

File: Directory.Build.targets (hunk 7 above)
Comment (weltkante) [reply Directory.Build.targets]: @RussKie [looks like](https://github.com/xunit/visualstudio.xunit/issues/148#issuecomment-579156956) this was a bug in VS which has recently been fixed in 16.4 but you are still referencing `Microsoft.Net.Test.Sdk` in version 16.1.1 which as far as I understand gets pulled in by the .NET Core SDK. Unfortunately referencing the nuget package directly from the test project gets me an error

> The framework 'Microsoft.NETCore.App', version '5.0.0-alpha1.19514.1' was not found.

which happens because the winforms repo `build-local.ps1` only installs a 64 bit SDK. Installing the SDK [manually](https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.0-alpha1-19514/dotnet-sdk-5.0.0-alpha1-19514-win-x64.exe) makes both 32bit and 64bit tests work within VS.

I think your build infrastructure already supports installing both SDKs, the aspnetcore repo seems to install both bitness in its build scripts, found [this log](https://github.com/dotnet/aspnetcore/issues/16664#issuecomment-553601518) while searching for the download link, but I have no idea how to configure the build scripts.

in summary:
* .NET Core SDK needs to be updated to reference `Microsoft.Net.Test.Sdk` in version 16.4+ instead of 16.1.1
  * alternatively somehow override the nuget package which gets pulled in by the SDK
* you need to install both x86 and x64 SDKs in your build scripts (like aspnetcore does?)

Created issue #2927 for tracking the problem

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs (hunk 5 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs]: Thank you for checking.

@JeremyKuhne any recommenadions?

I'll look into it on Monday

File: src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs (hunk 9 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs]: printDialog?

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs]: Yes, you need to `GC.KeepAlive` the `new` object. One gotcha that got me many years ago is that collection behavior is different in debug versus release. When running a debug build your locals are scoped to the end of the frame (the method or it's parent method  if it was inlined). I don't know precisely how the heuristics work and if there are any differences in Core. I'll try and make a point of finding out exactly how the frame GC reporting works, but depending on the implementation isn't the way to go. We need to go by the specification, which is as soon as you're no longer directly referencing a managed heap pointer, it's eligible for collection.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintDialog.cs]: Updated
